hubble: from and to cluster filters

kaworu opened the pull request:
https://github.com/cilium/cilium/pull/32313 introduced cluster name info at the Hubble flow Endpoint level. This PR adds corresponding filters on the Hubble server side, and two new `--from-cluster` and `--to-cluster` Hubble CLI flags leveraging the new filter.

For simplicity's sake the filter is a strict match for now, with the possibility to extend if/when the need arise (e.g. to RE2).

kaworu committed:
hubble: add cluster name filter to FlowFilter protobuf definition

Reword the IP.source_xlated documentation comment on the way.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

kaworu committed:
hubble: implement a new server-side cluster name filter

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

kaworu committed:
hubble/cli: implement --from-cluster and --to-cluster

Signed-off-by: Alexandre Perrin <alex@isovalent.com>

glibsm commented on pkg/hubble/filters/cluster_name.go:
```diff
@@ -0,0 +1,65 @@
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+)
+
+func sourceClusterName(ev *v1.Event) string {
+	return ev.GetFlow().GetSource().GetClusterName()
+}
+
+func destinationClusterName(ev *v1.Event) string {
+	return ev.GetFlow().GetDestination().GetClusterName()
+}
```
minor nit, but I don't think these intermediate functions are necessary.

kaworu replied:
Theses were "inspired" by the [`sourceIP`/`destinationIP`](https://github.com/cilium/cilium/blob/e49c7b3ddaf7e53a91a722ce7a1a328230749afc/pkg/hubble/filters/ip.go#L16-L22) but happy to change if there is a cleaner suggestion.

gandro commented on hubble/cmd/observe/flows_filter.go:
```diff
@@ -700,11 +698,19 @@ func (of *flowFilter) set(f *filterTracker, name, val string, track bool) error
+	case "to-cluster":
+		f.apply(func(f *flowpb.FlowFilter) {
+			f.DestinationClusterName = append(f.GetDestinationClusterName(), val)
+		})
```
Hm. So this is interesting: `--cluster` checks the node name, where as `--from/to-cluster` checks the endpoint data. The two have different semantics.

With this implementation, `--cluster foo` will only give you flows observed by nodes in foo. `--from-cluster foo || --to-cluster foo` however gives you flows going between the clusters, including flows observed by nodes in other clusters than foo (but going or coming from cluster foo).

This is different from other flags. For example, if one uses `--namespace bar` it is equivalent with `--from-namespace foo || --to-namespace foo`.

I think we should do a breaking change here and re-define how `--cluster` works, so it is consistent with all other flags. The old behavior is still available via `--node-name foo/`. We can document that.

kaworu replied:
@gandro yes all make sense to me, I was trying to avoid breaking as part of this PR but you are correct in that it's bringing inconsistency. Related question: do we start to document Hubble CLI changes in the Cilium upgrade guide?

gandro replied:
> Related question: do we start to document Hubble CLI changes in the Cilium upgrade guide?

Yeah, I think adding a new section to the upgrade notes makes sense to me. We probably also want to add it to the notes on the Hubble release in `cilium/hubble`, but than needs to be done by the release managers once Hubble CLI 1.16 ships

kaworu replied:
Thanks for the input, I've added a commit on top refactoring the `--cluster` flag as discussed, adding a note in the upgrade guide and (hopefully) summarized this discussion in the commit message. Please take another look :pray:

glrf replied:
I'm a bit torn on this. 

I agree that this new behaviour makes more sense and I'm oke with redefining how `--cluster` works.
What I don't particularly like is that `--cluster` will just stop working if you use the newly released Hubble CLI on an older cluster that doesn't support the new filter..

I don't have great idea to avoid this and I think in the end it makes more sense to have the flag this way. So I'm fine with keeping it as is, we just need to be aware of that.

chancez replied:
I've suggested we introduce some kind of version indicator previously. Eg: we could send a gRPC metadata header to clients when they connect, and the clients could use that to determine how they should behave.

kaworu replied:
@glrf good point, I agree with @chancez about having a version check as it's not the first time (nor the last) we encounter that kind of issue. @chancez do we have an issue for this already?

chancez replied:
I haven't since we discussed it in person and deemed it unnecessary/extra complexity in those cases at least.

kaworu replied:
Created https://github.com/cilium/cilium/issues/35724 to keep track of this with @chancez suggestion of sending the server version through gRPC metadata. Let me know if I missed something :pray:

kaworu committed:
hubble/cli: implement --cluster based on endpoint cluster name info

Before this patch, the Hubble CLI `--cluster` flag was implemented as
`--node-name cluster/`, i.e. filtering flows emitted by all the nodes in
the target cluster.

This behavior was inconsistent with the `--namespace foo` flag, which
shows not only flows ingressing/egressing the `foo` namespace, but also
flows observed from other namespaces either going to or coming from the
`foo` namespace. It also made the `--cluster` and `--node-name` flags
incompatible.

163c874f0b67 ("hubble: add cluster name to a flow's endpoints") added
cluster name info at the flow endpoint level, and the previous commits
added corresponding Hubble filters (`--from-cluster` and `--to-cluster`
in the Hubble CLI).

This patch refactors the `--cluster` flag to filter based on the newly
introduced endpoint cluster name info, making `--cluster foo`
conceptually equivalent to "--from-cluster foo || --to-namespace foo"
(and thus consistent with the `--namespace` flag) and removing the
incompatibility with the `--node-name` flag.

Signed-off-by: Alexandre Perrin <alex@isovalent.com>
Suggested-by: Sebastian Wicki <sebastian@isovalent.com>

glrf commented on Documentation/operations/upgrade.rst:
```diff
@@ -468,6 +468,17 @@ Changed Metrics
+Hubble CLI
+~~~~~~~~~~
+
+* the ``--cluster`` behavior changed to show flows emitted from nodes outside of
+  the provided cluster name (either coming from or going to the target cluster).
+  This change brings consistency between the ``--cluster`` and ``--namespace``
+  flags and removed the incompatibility between the ``--cluster`` and
+  ``--node-name`` flags. The previous behavior of ``--cluster foo`` can be
+  reproduced with ``--node-name foo/`` (shows all flows emitted from a node in
+  cluster ``foo``).
```
Can we make it even more clear that if you use the new Hubble CLI with an older Cilium cluster, that the `--cluster` flag won't work anymore? I don't think that's obvious if you read this and don't know how it's implemented.

gandro replied:
I'm surprised by the fact that the `Endpoint.Clustername` is only populated with the name of remote clusters, not the local cluster. Is that true?

I would have expected the Hubble parser to also populate the `ClusterName` field to also be set for local clusters, meaning the new version of the flag returns more flows than the old one, not fewer.

kaworu replied:
> I'm surprised by the fact that the Endpoint.Clustername is only populated with the name of remote clusters, not the local cluster. Is that true?

I can indeed repro with a local kind cluster using cilium `latest`: 

```json
    "destination": {
      "ID": 1305,
      "identity": 19193,
      "namespace": "kube-system",
      "labels": [
        "k8s:app.kubernetes.io/name=hubble-relay",
        "k8s:app.kubernetes.io/part-of=cilium",
        "k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=kube-system",
        "k8s:io.cilium.k8s.policy.cluster=default",
        "k8s:io.cilium.k8s.policy.serviceaccount=hubble-relay",
        "k8s:io.kubernetes.pod.namespace=kube-system",
        "k8s:k8s-app=hubble-relay"
      ],
      "pod_name": "hubble-relay-6c49d69996-xcq6g",
      "workloads": [
        {
          "name": "hubble-relay",
          "kind": "Deployment"
        }
      ]
    },
```
I checked to be sure that we have the endpoint cluster name in the server's protobuf definition
```
% grpcurl -plaintext localhost:4245  describe .flow.Endpoint
flow.Endpoint is a message:
message Endpoint {
  uint32 ID = 1;
  uint32 identity = 2;
  string namespace = 3;
  repeated string labels = 4;
  string pod_name = 5;
  repeated .flow.Workload workloads = 6;
  string cluster_name = 7;
}
```
So https://github.com/cilium/cilium/pull/32313 is included but not working as expected as we see `k8s:io.cilium.k8s.policy.cluster=default` in the labels but no endpoint's `cluster_name` field cc @rolinh. I took a quick look at the code but didn't find why it's not working.

kaworu replied:
@gandro ok so after further investigation:
> I'm surprised by the fact that the `Endpoint.Clustername` is only populated with the name of remote clusters, not the local cluster. Is that true?

The cluster name should be set regardless of whether the pod is from the local or a remote cluster. With that said, there is currently a bug where cluster name is not set for local endpoints (local to the node), that is why in my last comment attempt the cluster name is not set for Hubble Relay. https://github.com/cilium/cilium/pull/35415 should fix the missing cluster name for local endpoints (amongst other cases).

> I would have expected the Hubble parser to also populate the ClusterName field to also be set for local clusters, meaning the new version of the flag returns more flows than the old one, not fewer.

Conceptually yes, after the patch `--cluster foo` will show flows emitted by nodes outside of cluster `foo` either going to or from cluster `foo`, which `--node-name foo/` doesn't return. Technically, since the patch is based on endpoint labels, the new version will not show reserved identity flows (i.e. health check, node to node etc.). So the new set of returned flows is not a superset of the current version, but filtering out reserved identities flows is also consistent with the namespace filters. Does that make sense?

gandro replied:
Thanks for the in-depth explanation (and apologies for the delay in the response, this somehow slipped in my notifications)! Yes, that makes sense to me!

gandro reviewed (approved):
Thank you!
